src/test/ui/const-generics/fn-const-param-call.rs

```diff
@@ -9,12 +9,12 @@ fn function() -> u32 {
 
 struct Wrapper<const F: fn() -> u32>;
 
-impl<const F: fn() -> u32> Wrapper<{F}> {
+impl<const F: fn() -> u32> Wrapper<F> {
     fn call() -> u32 {
         F()
     }
 }
 
 fn main() {
-    assert_eq!(Wrapper::<{function}>::call(), 17);
+    assert_eq!(Wrapper::<function>::call(), 17);
```

Comment:
Oh wow this is really cool (even if I'm not sure it will survive `structural_match` requirements).

Reply:
Yeah, I am not worrying too much about this test as I also have a feeling most of this won't survive putting in place the `structural_match` requirement. As a side note, I was looking at implementing that requirement soon&#8482;.

Reply:
That was implemented in https://github.com/rust-lang/rust/pull/6562. Function types count as `structural_match`, though, as far as I can tell.

Reply:
@varkor I'm not seeing a reference to a PR in that issue. Can you point to the PR that implemented the `structural_match` constraint?

Reply:
(Sorry, my phone keyboard likes deleting the last character of pasted links for some reason.)

Reply:
Hang on, `ty::FnPtr` is considered `#[structural_match]`? Are raw pointers as well?

I would understand `ty::FnDef` but this right here is a function *pointer*, isn't it?

cc @pnkfelix

Reply:
The [RFC](https://github.com/rust-lang/rfcs/blob/master/text/1445-restrict-constants-in-patterns.md) doesn't even list the types that are intended to be `#[structural_match]`: it just says "builtin types", which would imply function pointers and raw pointers. Certainly that seems to be how `#[structural_match]` is currently implemented: it only really restricts pattern matching on user-defined types.